### PR TITLE
wip: attempt to use git LFS

### PR DESCRIPTION
**Pull request by paulirish**
I dropped `core/test/fixtures/traces/` and reintroduced it as `core/test/fixtures/lfs-traces/` but in LFS.

https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage
https://docs.github.com/en/repositories/working-with-files/managing-large-files/collaboration-with-git-large-file-storage

i'm curious how painful this is gonna be. as a `brew install` is needed.. i'm not sure what happens when folks without git-lfs try to use this branch. (and the potential future repo)
@brendankenny wanna try checking out this branch locally?

### CODE REVIEW

**Comment by connorjclark** [comment]
If you don't have git lfs installed, the files are simply a few lines of text referencing an external resource. Tests reading this would fail. At most, we would want to add a beforeAll thingy that checks git-lfs is installed (or just tries to read one of these files and check if they start with a `{`)

File: .gitattributes [hunk 1]
```diff
@@ -1,3 +1,9 @@
 *.js eol=lf
+
+# Always show in github ctrl-t file browser
 **/locales/*.json linguist-generated=true
-build/** linguist-generated=false
\ No newline at end of file
+build/** linguist-generated=false
+
+# Git large file storage
+*.psd filter=lfs diff=lfs merge=lfs -text
```
Comment (connorjclark) [inline .gitattributes:8]: remove the psd line? or am I overlooking a file in this PR...

**Comment by brendankenny** [comment]
I can try this out tomorrow, but what's the pitch for this for non-binary files? I don't like 1MiB traces in our repo as much as anyone, but they are text, so they diff pretty well.

I guess if we do automatic updates for the test traces that's potentially a high percentage of file churn in the diff * size of each file, so this would reduce the growth of the repo in that way? What's the workflow for reviewing new/changed traces? Assume they're test assets and not for human review, hope the tests are good?

**Comment by brendankenny** [comment]
> One very important thing to note is that checking out these files in CI count towards the LFS quota, which will add up fast. the action [nschloe/action-cached-lfs-checkout@v1](https://github.com/nschloe/action-cached-lfs-checkout) solves this problem nicely, though.

> [GitHub applies some fees when the total amount of LFS downloads surpasses 1GB.](https://github.com/github/roadmap/issues/237) Unfortunately, even gh-actions runs, when used with
> ```
> - name: Checkout code
>   uses: actions/checkout@v2
>   with:
>     lfs: true
> ```
> count towards that limit. To cache the LFS downloads, you can instead use this action. Simply replace the above by
> ```
> - name: Checkout code
>   uses: nschloe/action-cached-lfs-checkout@v1
>   # Use these to explicitly include/exclude files:
>   # with:
>   #   include: "*"
>   #   exclude: ""
> ```

The GHA cache is relatively long lived, seems kind of broken on GH's part that there's not some automatic version of this :/

**Comment by connorjclark** [comment]
I guess if we do automatic updates for the test traces that's potentially a high percentage of file churn in the diff * size of each file, so this would reduce the growth of the repo in that way?

That's the benefit as I understand it.

> What's the workflow for reviewing new/changed traces? Assume they're test assets and not for human review, hope the tests are good?

Unfortunately the GitHub ui won't show diffs for LFS objects, but you can at least get local diffing to treat these files like text by setting `git config diff.lfs.textconv cat`

> The GHA cache is relatively long lived, seems kind of broken on GH's part that there's not some automatic version of this :/

It's incredibly janky, but if they insist on charging for intra-network the same as out-of-network, that's on them.

**Comment by paulirish** [comment]
> I guess if we do automatic updates for the test traces that's potentially a high percentage of file churn in the diff * size of each file, so this would reduce the growth of the repo in that way? 

Yeah it means the repo is smaller. It still uses storage and bw. https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-storage-and-bandwidth-usage

But the main point is not having a repo that's enormous. right now [`git-sizer`](https://github.com/github/git-sizer) sez the repo has 2.7gb of blobs. that seems nonproblematic but i worry when we have the auto-trace thing going. 

> What's the workflow for reviewing new/changed traces? Assume they're test assets and not for human review, hope the tests are good?

yup. i don't think they've ever been reviewable. (but we do rely on them being parsable when we're debuggging)


---------

> action-cached-lfs-checkout

thats dope.

------

maybe we should do some rough estimations on the kind of trace file churn -> size increase we'd expect.

**Comment by paulirish** [comment]
<details>
<summary>current `git-sizer` output</summary>

```
❯ git-sizer --verbose
Processing blobs: 49438
Processing trees: 59036
Processing commits: 9724
Matching commits to trees: 9724
Processing annotated tags: 82
Processing references: 538
| Name                         | Value     | Level of concern               |
| ---------------------------- | --------- | ------------------------------ |
| Overall repository size      |           |                                |
| * Commits                    |           |                                |
|   * Count                    |  9.72 k   |                                |
|   * Total size               |  3.75 MiB |                                |
| * Trees                      |           |                                |
|   * Count                    |  59.0 k   |                                |
|   * Total size               |  37.6 MiB |                                |
|   * Total tree entries       |   951 k   |                                |
| * Blobs                      |           |                                |
|   * Count                    |  49.4 k   |                                |
|   * Total size               |  2.53 GiB |                                |
| * Annotated tags             |           |                                |
|   * Count                    |    82     |                                |
| * References                 |           |                                |
|   * Count                    |   538     |                                |
|     * Branches               |     1     |                                |
|     * Tags                   |   119     |                                |
|     * Remote-tracking refs   |   418     |                                |
|                              |           |                                |
| Biggest objects              |           |                                |
| * Commits                    |           |                                |
|   * Maximum size         [1] |  4.03 KiB |                                |
|   * Maximum parents      [2] |     2     |                                |
| * Trees                      |           |                                |
|   * Maximum entries      [3] |    63     |                                |
| * Blobs                      |           |                                |
|   * Maximum size         [4] |  32.1 MiB | ***                            |
|                              |           |                                |
| History structure            |           |                                |
| * Maximum history depth      |  5.16 k   |                                |
| * Maximum tag depth      [5] |     1     |                                |
|                              |           |                                |
| Biggest checkouts            |           |                                |
| * Number of directories  [6] |   247     |                                |
| * Maximum path depth     [6] |    10     | *                              |
| * Maximum path length    [6] |   132 B   | *                              |
| * Number of files        [7] |  1.55 k   |                                |
| * Total size of files    [8] |   114 MiB |                                |
| * Number of symlinks     [9] |     1     |                                |
| * Number of submodules       |     0     |                                |

[1]  c5cfac2b220b9f17a72cb3599a606adaf55820d7
[2]  aa1060077134f8785b62a6599660b132114cf5d5 (refs/remotes/origin/nativeheadlesssmoke)
[3]  bebc420aafb1946e814415d6169381ccd5d8176c (refs/remotes/origin/prioritize-lcp-image:lighthouse-core/audits)
[4]  40fce95e44d151a651a5e1ba767e2c4bb2c6a20e (refs/remotes/origin/branch-9:lighthouse-core/test/fixtures/fraggle-rock/artifacts/sample-flow-artifacts.json)
[5]  4ecfccb2f5e605e9740b1e316b88d481944305cb (refs/tags/1.0.6)
[6]  586874dda4f0934a93e9a6f601488631e12f51d0 (03af4038e0f0168c8d93b8706e8d287592704008^{tree})
[7]  a760f980562470ce97f8a439ee734506d8c1f14e (refs/remotes/origin/esm-core^{tree})
[8]  95b1b8e8af172319962781b2a5554cebe5e89775 (refs/remotes/origin/injected-iframes-2^{tree})
[9]  79aff1417ba014da4a8f763af10876caba4b25b0 (7e3912d40bea6382389c768a98cf2106b5feb76e^{tree})

```

</details>


basically... not a lot of concern. :)

heh.. meanwhile here's chromium...


<details>
<summary>chromium repo w/ `git-sizer`</summary>

```
~/chromium/src 
❯ git-sizer --verbose
Processing blobs: 7407851
Processing trees: 8643249
Processing commits: 1236130
Matching commits to trees: 1236130
Processing annotated tags: 0
Processing references: 16834
| Name                         | Value     | Level of concern               |
| ---------------------------- | --------- | ------------------------------ |
| Overall repository size      |           |                                |
| * Commits                    |           |                                |
|   * Count                    |  1.24 M   | **                             |
|   * Total size               |  1.03 GiB | ****                           |
| * Trees                      |           |                                |
|   * Count                    |  8.64 M   | *****                          |
|   * Total size               |  28.0 GiB | ***************                |
|   * Total tree entries       |   648 M   | ************                   |
| * Blobs                      |           |                                |
|   * Count                    |  7.41 M   | ****                           |
|   * Total size               |   727 GiB | !!!!!!!!!!!!!!!!!!!!!!!!!!!!!! |
| * Annotated tags             |           |                                |
|   * Count                    |     0     |                                |
| * References                 |           |                                |
|   * Count                    |  16.8 k   |                                |
|     * Branches               |    27     |                                |
|     * Tags                   |  13.0 k   |                                |
|     * Remote-tracking refs   |  3.77 k   |                                |
|     * Git stash              |     1     |                                |
|     * Other                  |     2     |                                |
|                              |           |                                |
| Biggest objects              |           |                                |
| * Commits                    |           |                                |
|   * Maximum size         [1] |   392 KiB | ********                       |
|   * Maximum parents      [2] |     2     |                                |
| * Trees                      |           |                                |
|   * Maximum entries      [3] |  2.62 k   | **                             |
| * Blobs                      |           |                                |
|   * Maximum size         [4] |   508 MiB | !!!!!!!!!!!!!!!!!!!!!!!!!!!!!! |
|                              |           |                                |
| History structure            |           |                                |
| * Maximum history depth      |   915 k   | *                              |
| * Maximum tag depth          |     0     |                                |
|                              |           |                                |
| Biggest checkouts            |           |                                |
| * Number of directories  [5] |  32.7 k   | ****************               |
| * Maximum path depth     [6] |    18     | *                              |
| * Maximum path length    [7] |   284 B   | **                             |
| * Number of files        [8] |   390 k   | *******                        |
| * Total size of files    [8] |  3.30 GiB | ***                            |
| * Number of symlinks     [9] |    13     |                                |
| * Number of submodules  [10] |   114     | *                              |
```

</details>

File: .gitattributes [hunk 2]
```diff
@@ -5,5 +5,4 @@
 build/** linguist-generated=false
 
 # Git large file storage
-*.psd filter=lfs diff=lfs merge=lfs -text
-**/fixtures/lfs-traces/** filter=lfs diff=lfs merge=lfs -text
+**/fixtures/lfs-traces/** filter=lfs diff=lfs merge=lfs
```
Comment (connorjclark) [inline .gitattributes:8]: Good call removing -text... I think? I'm guessing here, but does this change mean that git will never attempt to auto-resolve a conflict on these files, right (treat it like binary?)

Comment (paulirish) [reply .gitattributes:8]: shrug... i read this:

>  finally the -text argument tells Git that this is not a text file, which is a strange way to tell it that it's a binary file.

but from a not totally reliable source. 

a nice source i did find was https://github.com/git-lfs/git-lfs/issues/4881#issuecomment-1041581286

Comment (connorjclark) [reply .gitattributes:8]: if they are adding it by default, seems we should keep it.

**Comment by paulirish** [comment]
In earlier discussion we decided to pursue #14525 first. Maybe we'll do this but... probably not.

**Pull request closed without merging**